devices/src/virtio/net.rs

```diff
@@ -868,17 +905,6 @@ impl VirtioDevice for Net {
 
             //TODO: barrier needed here maybe?
 
-            epoll::ctl(
-                self.epoll_config.epoll_raw_fd,
-                epoll::ControlOptions::EPOLL_CTL_ADD,
-                tap_raw_fd,
-                epoll::Event::new(epoll::Events::EPOLLIN, self.epoll_config.rx_tap_token),
```

Comment:
I see this is no longer registered in `activate()`, but on the `RX_QUEUE` event. Are we sure the driver sends these events on initialization as well? It's my understanding that these events are issued by the driver when it has processed a/some descriptior(s) from the RX queue and there is now room for more.

If we never put anything there, does the spec say that the driver should notify regardless?

Reply:
The virt queue buffers are set up after device activation time. Actually, "activation" is a Firecracker-defined term, not mentioned in the virtio spec. We use it to refer to the process we need to go through after device initialization or reset (after the driver sets the DRIVER_OK bit - see section 3.1.1 in the virtio spec, which refers to the device as being "live," after this step).

Populating the virt queues is a subsequent step, performed by the driver, after the device is "live." While I found no "must notify" constraint on the driver in the spec, section 2.6.13 describes the queue hydration procedure, for which the final step is:

> 7. The driver sends an available buffer notification to the device if such notifications are not suppressed.

All in all, IMO we can be as certain as the spec is that, whenever the driver supplies an available buffer (or a batch thereof), we'll get an RX_QUEUE event.